File: packages/esm-format-go/pkg/esm/reservoir_species_roundtrip_test.go
package esm

// Round-trip coverage for Species.constant (reservoir species) — gt-ertm.

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestReservoirSpeciesConstantRoundTrip(t *testing.T) {
	wd, err := os.Getwd()
	require.NoError(t, err)
	repoRoot := filepath.Join(wd, "..", "..", "..", "..")
	path := filepath.Join(repoRoot, "tests", "valid", "reservoir_species_constant.esm")

	parsed, err := Load(path)
	require.NoError(t, err)
	require.NotNil(t, parsed)

	rs, ok := parsed.ReactionSystems["SuperFastSubset"]
	require.True(t, ok)

	for _, name := range []string{"O2", "CH4", "H2O"} {
		sp, ok := rs.Species[name]
		require.True(t, ok, "species %s missing", name)
		require.NotNil(t, sp.Constant, "species %s should have constant flag", name)
		assert.True(t, *sp.Constant, "species %s should be constant=true", name)
	}
	for _, name := range []string{"O3", "OH", "HO2"} {
		sp, ok := rs.Species[name]
		require.True(t, ok, "species %s missing", name)
		assert.Nil(t, sp.Constant, "species %s should have no constant flag", name)
	}

	first, err := json.Marshal(parsed)
	require.NoError(t, err)
	var reparsed EsmFile
	require.NoError(t, json.Unmarshal(first, &reparsed))
	second, err := json.Marshal(&reparsed)
	require.NoError(t, err)
	assert.JSONEq(t, string(first), string(second))
}

File: packages/earthsci-toolkit/src/reservoir-species-roundtrip.test.ts
/**
 * Round-trip coverage for Species.constant (reservoir species) — gt-ertm.
 *
 * Loads a reaction system with three reservoir species (O2, CH4, H2O) declared
 * via constant=true, plus three ordinary state species. Verifies parse →
 * serialize → reparse is JSON-byte-equivalent and that the constant flag is
 * only present on the reservoir species.
 */
import { describe, it, expect } from 'vitest'
import { readFileSync } from 'fs'
import { join } from 'path'
import { load, save } from './index.js'

const FIXTURE = join(__dirname, '../../../tests/valid/reservoir_species_constant.esm')

describe('reservoir species (Species.constant=true)', () => {
  it('round-trips the constant flag byte-identical and flags only reservoir species', () => {
    const raw = readFileSync(FIXTURE, 'utf-8')
    const parsed = load(raw) as Record<string, unknown>
    const rs = (parsed.reaction_systems as Record<string, { species: Record<string, { constant?: boolean }> }>)
      .SuperFastSubset
    for (const name of ['O2', 'CH4', 'H2O']) {
      expect(rs.species[name].constant).toBe(true)
    }
    for (const name of ['O3', 'OH', 'HO2']) {
      expect(rs.species[name].constant).toBeUndefined()
    }

    const serialized = save(parsed as never)
    const original = JSON.parse(raw)
    const reserialized = JSON.parse(serialized)
    expect(reserialized).toEqual(original)
  })
})
